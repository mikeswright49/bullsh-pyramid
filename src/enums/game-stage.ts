export enum GameStage {
    Initiation,
    Memorization,
    Flipping,
    Declaration,
    Bullshit,
    Memory,
    Complete,
    Reveal,
}
