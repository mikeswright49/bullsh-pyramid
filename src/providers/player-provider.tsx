import React from 'react';
import { useState, useEffect } from 'react';
import { Player } from 'types/player';
import { PlayerStore } from 'src/stores/player-store';
export const PlayerContext = React.createContext<Player>(null);

export function PlayerProvider(props: React.PropsWithChildren<{ playerId: string }>) {
    const [player, setPlayer] = useState<Player>();

    useEffect(() => {
        if (!props.playerId) {
            return;
        }
        PlayerStore.subscribeToPlayer(props.playerId, (updatedPlayer) => {
            setPlayer(updatedPlayer);
        });

        return () => PlayerStore.unsubscribeToPlayer();
    }, [props.playerId]);

    return <PlayerContext.Provider value={player}>{props.children}</PlayerContext.Provider>;
}
