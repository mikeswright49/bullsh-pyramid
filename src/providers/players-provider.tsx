import React from 'react';
import { useState, useEffect } from 'react';
import { Player } from 'types/player';
import { GameStore } from 'src/stores/game-store';

export const PlayersContext = React.createContext<Player[]>([]);

export function PlayersProvider(props: React.PropsWithChildren<{ gameId: string }>) {
    const [players, setPlayers] = useState({});
    useEffect(() => {
        GameStore.subscribeToPlayers(props.gameId, (player: Player) => {
            players[player.id] = player;
            setPlayers({ ...players });
        });
    }, [props.gameId]);
    return (
        <PlayersContext.Provider value={Object.values(players)}>
            {props.children}
        </PlayersContext.Provider>
    );
}
