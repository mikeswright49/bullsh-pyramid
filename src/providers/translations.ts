export const TRANSLATIONS = {
    en: {
        'game.board.current.row': 'Current Row: {{ activeRow }}',
        'game.board.current.card': 'Current card',
        'game.display.stage.memorization.text': 'You have 30s to remember your cards!',
        'game.display.stage.flipping.text': 'The host is flipping a card',
        'game.display.stage.declaration.text': 'Let the world know if you have it',
        'game.display.stage.assign.text': 'Players are beginning to lie left right and center',
        'game.display.stage.bullshit.text': 'People should know who you think was lieing',
        'game.display.stage.reveal.text': 'We all now know the truth',
        'game.display.stage.memory.text': 'Well which cards did you have?',
        'game.display.stage.complete.text': 'Well time for a new game',
    },
};
