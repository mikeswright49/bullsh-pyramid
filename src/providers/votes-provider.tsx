import React from 'react';
import { useState, useEffect } from 'react';
import { Vote } from 'types/vote';
import { VoteStore } from 'src/stores/vote-store';
export const VotesContext = React.createContext<Vote[]>([]);

export function VotesProvider(props: React.PropsWithChildren<{ gameId: string }>) {
    const [votes, setVotes] = useState([]);
    useEffect(() => {
        if (!votes) {
            return;
        }

        VoteStore.subscribeToVotes(props.gameId, (vote: Vote) => {
            votes[vote.id] = vote;
            setVotes({ ...votes });
        });
    }, [props.gameId]);

    return (
        <VotesContext.Provider value={Object.values(votes)}>{props.children}</VotesContext.Provider>
    );
}
