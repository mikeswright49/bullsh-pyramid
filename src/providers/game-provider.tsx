import React, { useEffect, useState } from 'react';
import { GameState } from 'types/game-state';
import { GameStore } from 'src/stores/game-store';
import { GameStage } from 'src/enums/game-stage';

export const DEFAULT_GAME_STATE: GameState = {
    id: '',
    gameStage: GameStage.Initiation,
    activeRow: 0,
    activeIndex: 0,
    cardCount: 4,
    flipDelay: 5000,
    players: [],
    tiers: [],
};
export const GameContext = React.createContext<GameState>(DEFAULT_GAME_STATE);

export function GameProvider(props: React.PropsWithChildren<{ gameId: string }>) {
    const { gameId } = props;
    const [gameState, setGameState] = useState<GameState>(DEFAULT_GAME_STATE);

    useEffect(() => {
        if (!gameId) {
            return;
        }

        GameStore.subscribeToGame(gameId, (updatedGameState) => {
            setGameState(updatedGameState);
        });

        return () => GameStore.unsubscribeToGame();
    }, [gameId]);

    return <GameContext.Provider value={gameState}>{props.children}</GameContext.Provider>;
}
