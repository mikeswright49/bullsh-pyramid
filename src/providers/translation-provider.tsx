import React, { useContext } from 'react';
import { useEffect, useState } from 'react';
import { GameContext } from './game-provider';
import { TRANSLATIONS } from '../translations/translations';

const INTERPOLATE_REGEX = /{{.*?}}/gi;
const INTERPOLATE_PROPERTY_REGEX = /{{(.*)?}}/i;
const INTERPOLATE_CAPTURE_COUNT = 2;

export type Translations = {
    translations: { [key: string]: string };
    translate: (key: string, variables?: { [key: string]: string | number | boolean }) => string;
};

export const TranslationContext = React.createContext<Translations>({
    translate: (a) => a,
    translations: {},
});

export function TranslationProvider(props: React.PropsWithChildren<unknown>) {
    const { language } = useContext(GameContext);
    const [translations, setTranslations] = useState({});
    useEffect(() => {
        setTranslations(TRANSLATIONS[language || 'en']);
    }, [language]);

    /**
     * Returns the nested property value (userMetadata.giftcard.bacon)
     * @param source property value
     */
    function getProperty(source: string[], parent: unknown): string {
        if (!source || !parent) {
            return;
        }
        if (source.length > 1) {
            const property: string = source.shift();
            return this.getProperty(source, parent[property.trim()]);
        }
        return parent[source[0].trim()];
    }

    /**
     * Evaluates the {{}} interpolated strings
     * @param source Source content to be parsed
     * @param data Data object to read values off of
     */
    function evaluateInterpolate(
        source: string,
        data: { [key: string]: string | number | boolean }
    ): string {
        if (!data) {
            return source;
        }

        const matches = source.match(INTERPOLATE_REGEX) || [];
        matches.forEach((match: string) => {
            const properties = INTERPOLATE_PROPERTY_REGEX.exec(match);
            if (properties && properties.length === INTERPOLATE_CAPTURE_COUNT) {
                const property = getProperty(properties[1].split('.'), data);
                source = source.replace(match, property || '');
            }
        });
        return source;
    }

    function translate(key: string, variables?: { [key: string]: string | number | boolean }) {
        if (!translations) {
            return '';
        }
        return evaluateInterpolate(translations[key], variables);
    }

    return (
        <TranslationContext.Provider value={{ translations, translate }}>
            {props.children}
        </TranslationContext.Provider>
    );
}
