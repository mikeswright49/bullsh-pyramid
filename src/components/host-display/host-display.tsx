import React, { useContext, useEffect, useRef } from 'react';
import { GameContext } from 'src/providers/game-provider';

import { GameStage } from 'src/enums/game-stage';
import { PlayersContext } from 'src/providers/players-provider';
import { transitionToMemorization } from './transitions/to-memorization';
import { transitionToFlipping } from './transitions/to-flipping';
import { transitionToBullshit } from './transitions/to-bullshit';
import { transitionToNewGame } from './transitions/to-new-game';
import { transitionToDeclaration } from './transitions/to-declaration';
import { transitionToReveal } from './transitions/to-reveal';
import { transitionFromReveal } from './transitions/from-reveal';
import { transitionToAssignment } from './transitions/to-assignment';
import { TranslationContext } from 'src/providers/translation-provider';

const MEMORIZATION_TIMEOUT = 30000;
const MEMORY_TIMEOUT = 1000;

export function HostDisplay() {
    const gameState = useContext(GameContext);
    const players = useContext(PlayersContext);
    const { translate } = useContext(TranslationContext);
    const { gameStage, id: gameId, flipDelay } = gameState;
    const timeoutRef = useRef(null);

    /**
     * Transition to the flip phase of the game
     * This is some fucked up shit right here.
     * This is where hooks sucks
     */
    useEffect(() => {
        if (timeoutRef.current) {
            clearTimeout(timeoutRef.current);
        }

        if (gameStage === GameStage.Memorization) {
            timeoutRef.current = setTimeout(
                () => transitionToFlipping(players, gameState),
                MEMORIZATION_TIMEOUT
            );
        } else if (gameStage === GameStage.Declaration) {
            timeoutRef.current = setTimeout(() => transitionToAssignment(gameId), flipDelay);
        } else if (gameStage === GameStage.Memory) {
            timeoutRef.current = setTimeout(() => transitionToNewGame(gameId), MEMORY_TIMEOUT);
        }
        return () => {
            if (timeoutRef.current) {
                clearTimeout(timeoutRef.current);
            }
        };
    }, [gameStage]);

    function renderHostDisplay() {
        switch (gameStage) {
            case GameStage.Initiation:
                return (
                    <div className="stack-y-2">
                        <h2>{translate('host.display.stage.initiation.title')}</h2>
                        <button
                            className="pure-button pure-button-primary"
                            onClick={() => transitionToMemorization(players, gameState)}
                        >
                            {translate('host.display.stage.initiation.button')}
                        </button>
                    </div>
                );
            case GameStage.Memorization:
                return (
                    <div className="stack-y-2">
                        <h3>{translate('host.display.stage.memorization.title')}</h3>
                        <h4>{translate('host.display.stage.memorization.subtitle')}</h4>
                    </div>
                );
            case GameStage.Flipping:
                return (
                    <div className="stack-y-2">
                        <h3>{translate('host.display.stage.flipping.title')}</h3>
                        <button
                            className="pure-button pure-button-primary"
                            onClick={() => transitionToDeclaration(gameState)}
                        >
                            {translate('host.display.stage.flipping.button')}
                        </button>
                    </div>
                );
            case GameStage.Declaration:
                return (
                    <div className="stack-y-2">
                        <button
                            className="pure-button pure-button-primary"
                            onClick={() => {
                                clearTimeout(timeoutRef.current);
                                transitionToAssignment(gameState.id);
                            }}
                        >
                            {translate('host.display.stage.declaration.button')}
                        </button>
                    </div>
                );
            case GameStage.Assign:
                return (
                    <div className="stack-y-2">
                        <h3>{translate('host.display.stage.assign.title')}</h3>
                        <button
                            className="pure-button pure-button-primary"
                            onClick={() => transitionToBullshit(gameState.id)}
                        >
                            {translate('host.display.stage.assign.button')}
                        </button>
                    </div>
                );
            case GameStage.Bullshit:
                return (
                    <div className="stack-y-2">
                        <h3>{translate('host.display.stage.bullshit.title')}</h3>
                        <button
                            className="pure-button pure-button-primary"
                            onClick={() => transitionToReveal(gameState.id)}
                        >
                            {translate('host.display.stage.bullshit.button')}
                        </button>
                    </div>
                );
            case GameStage.Reveal:
                return (
                    <div className="stack-y-2">
                        <h3>{translate('host.display.stage.reveal.title')}</h3>
                        <button
                            className="pure-button pure-button-primary"
                            onClick={() => transitionFromReveal(players, gameState)}
                        >
                            {translate('host.display.stage.reveal.button')}
                        </button>
                    </div>
                );
            default:
                return null;
        }
    }

    if (!gameState) {
        return null;
    }

    return <div data-testid="host-display">{renderHostDisplay()}</div>;
}
