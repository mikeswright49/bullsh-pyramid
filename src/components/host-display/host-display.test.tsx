import React from 'react';
import { GameStage } from 'src/enums/game-stage';
import { HostDisplay } from './host-display';
import { GameContext } from 'src/providers/game-provider';
import { render, screen } from '@testing-library/react';
import { PlayersContext } from 'src/providers/players-provider';
import { GameState } from 'types/game-state';
import { MOCK_GAME_STATE } from 'src/__tests__/data/mock-game-state';
import { MOCK_PLAYERS } from 'src/__tests__/data/mock-players';
import { TranslationContext } from 'src/providers/translation-provider';
import { MOCK_TRANSLATIONS } from 'src/__tests__/data/mock-translations';

describe('HostDisplay', () => {
    let gameContextMock: GameState;
    beforeEach(() => {
        gameContextMock = Object.assign({}, MOCK_GAME_STATE);
    });

    const testCases = Object.keys(GameStage)
        .map((key) => (Number.isInteger(parseInt(key, 10)) ? null : [key]))
        .filter((x) => !!x);
    test.each(testCases)('it should be able to render the host view for %s', (stage) => {
        gameContextMock.gameStage = GameStage[stage];
        render(
            <GameContext.Provider value={gameContextMock}>
                <PlayersContext.Provider value={MOCK_PLAYERS}>
                    <TranslationContext.Provider value={MOCK_TRANSLATIONS}>
                        <HostDisplay />
                    </TranslationContext.Provider>
                </PlayersContext.Provider>
            </GameContext.Provider>
        );
        expect(screen.getByTestId('host-display')).toMatchSnapshot();
    });
});
