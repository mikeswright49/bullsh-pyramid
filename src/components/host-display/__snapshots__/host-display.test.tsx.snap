// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HostDisplay it should be able to render the host view for Assign 1`] = `
<div
  data-testid="host-display"
>
  <div
    class="stack-y-2"
  >
    <h3>
      host.display.stage.assign.title
    </h3>
    <button
      class="pure-button pure-button-primary"
    >
      host.display.stage.assign.button
    </button>
  </div>
</div>
`;

exports[`HostDisplay it should be able to render the host view for Bullshit 1`] = `
<div
  data-testid="host-display"
>
  <div
    class="stack-y-2"
  >
    <h3>
      host.display.stage.bullshit.title
    </h3>
    <button
      class="pure-button pure-button-primary"
    >
      host.display.stage.bullshit.button
    </button>
  </div>
</div>
`;

exports[`HostDisplay it should be able to render the host view for Complete 1`] = `
<div
  data-testid="host-display"
/>
`;

exports[`HostDisplay it should be able to render the host view for Declaration 1`] = `
<div
  data-testid="host-display"
>
  <div
    class="stack-y-2"
  >
    <button
      class="pure-button pure-button-primary"
    >
      host.display.stage.declaration.button
    </button>
  </div>
</div>
`;

exports[`HostDisplay it should be able to render the host view for Flipping 1`] = `
<div
  data-testid="host-display"
>
  <div
    class="stack-y-2"
  >
    <h3>
      host.display.stage.flipping.title
    </h3>
    <button
      class="pure-button pure-button-primary"
    >
      host.display.stage.flipping.button
    </button>
  </div>
</div>
`;

exports[`HostDisplay it should be able to render the host view for Initiation 1`] = `
<div
  data-testid="host-display"
>
  <div
    class="stack-y-2"
  >
    <h2>
      host.display.stage.initiation.title
    </h2>
    <button
      class="pure-button pure-button-primary"
    >
      host.display.stage.initiation.button
    </button>
  </div>
</div>
`;

exports[`HostDisplay it should be able to render the host view for Memorization 1`] = `
<div
  data-testid="host-display"
>
  <div
    class="stack-y-2"
  >
    <h3>
      host.display.stage.memorization.title
    </h3>
    <h4>
      host.display.stage.memorization.subtitle
    </h4>
  </div>
</div>
`;

exports[`HostDisplay it should be able to render the host view for Memory 1`] = `
<div
  data-testid="host-display"
/>
`;

exports[`HostDisplay it should be able to render the host view for Reveal 1`] = `
<div
  data-testid="host-display"
>
  <div
    class="stack-y-2"
  >
    <h3>
      host.display.stage.reveal.title
    </h3>
    <button
      class="pure-button pure-button-primary"
    >
      host.display.stage.reveal.button
    </button>
  </div>
</div>
`;
